engine-precompiles/src/alt_bn256.rs

```diff
@@ -42,21 +42,144 @@ mod consts {
 
     /// Pair element length.
     pub(super) const PAIR_ELEMENT_LEN: usize = 192;
+
+    pub(super) const SCALAR_PART_LEN: usize = SCALAR_LEN / 2;
+
+    /// Size of BN scalars.
+    pub(super) const SCALAR_LEN: usize = 32;
+
+    /// Half the size of a point size.
+    pub(super) const POINT_PART_LEN: usize = POINT_LEN / 2;
+
+    /// Size of BN points.
+    pub(super) const POINT_LEN: usize = 64;
+
+    /// Size of BN pairs.
+    pub(super) const POINT_PAIR_LEN: usize = 128;
+
+    /// Output length.
+    pub(super) const OUTPUT_LEN: usize = 64;
+}
+
+#[cfg(feature = "contract")]
+trait SdkEncode {
+    type Encoded;
+
+    fn sdk_encode(self) -> Self::Encoded;
+}
+
+#[cfg(feature = "contract")]
+impl SdkEncode for bn::Fr {
+    type Encoded = [u8; consts::SCALAR_LEN];
+
+    fn sdk_encode(self) -> Self::Encoded {
+        let (low, high) = {
+            let [low_int, high_int] = self.into_u256().0;
+            (low_int.to_le_bytes(), high_int.to_le_bytes())
+        };
+
+        let mut bytes = [0u8; consts::SCALAR_LEN];
+        bytes[0..consts::SCALAR_PART_LEN].copy_from_slice(&low);
+        bytes[consts::SCALAR_PART_LEN..consts::SCALAR_PART_LEN * 2].copy_from_slice(&high);
+        bytes
+    }
+}
+
+#[cfg(feature = "contract")]
+impl SdkEncode for bn::G1 {
+    type Encoded = [u8; consts::POINT_LEN];
+
+    fn sdk_encode(self) -> Self::Encoded {
+        bn::AffineG1::from_jacobian(self)
+            .map(|p| {
+                // It is possible to use sdk_encode for Fr here but, this is
+                // just a small optimisation to avoid additional allocations.
+                let (px, py) = (p.x(), p.y());
+                let (px_low, px_high) = {
+                    let [low_int, high_int] = px.into_u256().0;
+                    (low_int.to_le_bytes(), high_int.to_le_bytes())
+                };
+                let (py_low, py_high) = {
+                    let [low_int, high_int] = py.into_u256().0;
+                    (low_int.to_le_bytes(), high_int.to_le_bytes())
+                };
+
+                let mut bytes = [0u8; consts::POINT_LEN];
+                bytes[0..consts::SCALAR_PART_LEN].copy_from_slice(&px_low);
+                bytes[consts::SCALAR_PART_LEN..consts::SCALAR_PART_LEN * 2]
+                    .copy_from_slice(&px_high);
+                bytes[consts::SCALAR_PART_LEN * 2..consts::SCALAR_PART_LEN * 3]
+                    .copy_from_slice(&py_low);
+                bytes[consts::SCALAR_PART_LEN * 3..consts::SCALAR_PART_LEN * 4]
+                    .copy_from_slice(&py_high);
+                bytes
+            })
+            .unwrap_or_else(|| [0u8; consts::POINT_LEN])
+    }
+}
+
+#[cfg(feature = "contract")]
+impl SdkEncode for bn::G2 {
+    type Encoded = [u8; consts::POINT_PAIR_LEN];
+
+    fn sdk_encode(self) -> Self::Encoded {
```

Comment:
I wonder if there is a way to unify all 3 of these implementations. They have the same basic structure: turn the input into its component 256-bit numbers and then concatenate them all together. Maybe some cute const generic trick could work here where we write a function like `fn concat<N: usize>(xs: [U256; N]) -> [u8; 32 * N]`.

Reply:
I had thought about doing something along those lines, honestly. I think NEAR does something similar too.

Reply:
Done in https://github.com/aurora-is-near/aurora-engine/pull/540/commits/54faa982eeccff75ef468db27f6632f62c1decc7